File: blasty-back/src/main/java/com/blasty/service/Interface/ParkingService.java
package com.blasty.service.Interface;
import com.blasty.dto.request.ParkingRequest;
import com.blasty.dto.response.ParkingOccupancyResponse;
import com.blasty.dto.response.ParkingResponse;
import com.blasty.dto.response.ParkingRevenueResponse;

import java.util.List;

public interface ParkingService {
    ParkingResponse createParking(ParkingRequest request);

    ParkingResponse getParkingById(Long id);

    List<ParkingResponse> getAllParkings();

    ParkingResponse updateParking(Long id, ParkingRequest request);

    int getAvailablePlaces(Long parkingId);

    void deleteParking(Long id);

    ParkingOccupancyResponse getParkingOccupancy(Long id);

    List<ParkingRevenueResponse> getAllParkingsRevenue(String period);

    ParkingRevenueResponse getParkingRevenue(Long id , String period);

    List<ParkingOccupancyResponse> getAllParkingsOccupancy();

    void incrementOccupiedSpaces(Long parkingId);

    void decrementOccupiedSpaces(Long parkingId);
}

File: blasty-frontend/src/app/core/services/parking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { 
  Parking, 
  ParkingRequest, 
  ParkingResponse, 
  ParkingOccupancyResponse, 
  ParkingRevenueResponse 
} from '../models/parking.model';
import { environment } from '../../environment/environment';

@Injectable({
  providedIn: 'root'
})
export class ParkingService {
  private baseUrl = `${environment.apiURL}/parkings`;

  constructor(private http: HttpClient) { }

  // Admin endpoints
  createParking(request: ParkingRequest): Observable<ParkingResponse> {
    return this.http.post<ParkingResponse>(this.baseUrl, request);
  }

  updateParking(id: number, request: ParkingRequest): Observable<ParkingResponse> {
    return this.http.put<ParkingResponse>(`${this.baseUrl}/${id}`, request);
  }

  deleteParking(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getParkingOccupancy(id:number): Observable<ParkingOccupancyResponse[]> {
    return this.http.get<ParkingOccupancyResponse[]>(`${this.baseUrl}/${id}/occupancy`);
  }

  getParkingRevenue(id:number ,period: string = 'month'): Observable<ParkingRevenueResponse[]> {
    const params = new HttpParams().set('period', period);
    return this.http.get<ParkingRevenueResponse[]>(`${this.baseUrl}/${id}/revenue`, { params });
  }

  // Client accessible endpoints
  getParkingById(id: number): Observable<ParkingResponse> {
    return this.http.get<ParkingResponse>(`${this.baseUrl}/${id}`);
  }

  getAllParkings(): Observable<ParkingResponse[]> {
    return this.http.get<ParkingResponse[]>(this.baseUrl);
  }

  getAvailablePlaces(id: number): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/${id}/available-places`);
  }
}